### ISSUE TITLE
Search: Filter by "match similarity" threshold

### ISSUE THREAD

**Issue by NeilMBenetech** [benetech/ServiceNet#781]
On the December 2019 partner working group call, John Higgins was asked if, when we are seeking matches at the Organization level, we generate a numeric value reflecting how well matched a record is with another Organization record at any particular partner. John noted the various factors that contribute to this score, like name similarity, website address, etc and that those factors will change over time as we tune our algorithm. For now we'll call this numeric value the "match similarity score".

This task is to create a new search filter that allows users to only return results whose "match similarity score" exceeds a designated threshold. 

Rather than expose to users the actual "match similarity score" which could change radically over time as we change our algorithm, I propose the filter be driven by relative "match similarity scores" divided into, say, quartiles. So all matches between any two Organization records with be either in the 1st, 2nd, 3rd or 4th "relative match similarity" quartile compared with all other matches in the database. 

Then the search filter for the user would limit search results to just those whose "relative match similarity" is in the 4th/top quartile. Rather than name this filter something as strange as that wording, instead let's just create for now a single checkbox for this top quartile called "Only show highly-matched Organizations".

**Comment by johnhbenetech** [comment]
@kdondziak When attempting to log in with various non admin users, I get the following ISE which also prevents any records from displaying on the main screen:

```
Internal Server Error

java.lang.NullPointerException: null
        at org.benetech.servicenet.repository.ActivityRepository.addSimilarityFilter(ActivityRepository.java:230)
        at org.benetech.servicenet.repository.ActivityRepository.addFilters(ActivityRepository.java:385)
        at org.benetech.servicenet.repository.ActivityRepository.findAllWithFilters(ActivityRepository.java:126)
        at org.benetech.servicenet.repository.ActivityRepository$$FastClassBySpringCGLIB$$23b1a12b.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
```

**Comment by kdondziak** [comment]
@johnhbenetech Thanks for watchfulness. I will fix it immediately.

**Comment by bognasoldev89** [comment]
I tested it on match prepared by @kdondziak but I would like to wait for resolving matching calculations issues to test this issue properly.

**Comment by bognasoldev89** [comment]
This feature filter records correctly. I looked on randomly chosen records for different providers and I always got at least 75% match.

Filtering feature working:

![781 1](https://user-images.githubusercontent.com/51324797/74026923-7c2e5a00-49a7-11ea-9e03-81be8669d727.gif)

And some examples:

![781 2](https://user-images.githubusercontent.com/51324797/74026953-881a1c00-49a7-11ea-879d-2e2947b6c03b.png)

![781 3](https://user-images.githubusercontent.com/51324797/74026957-894b4900-49a7-11ea-8209-50b048b5b8ba.png)

![781 4](https://user-images.githubusercontent.com/51324797/74026959-894b4900-49a7-11ea-8264-2bc21f0a97c9.png)

![781 5](https://user-images.githubusercontent.com/51324797/74026963-89e3df80-49a7-11ea-837c-538de8ff36b1.png)

One thing though, this checkbox doesn't react to reset button on filtering feature. 

![781 bug](https://user-images.githubusercontent.com/51324797/74027109-cfa0a800-49a7-11ea-9e55-d190fc92004f.gif)

I assume, that it should be unchecked by default, that makes sense in my opinion. @johnhbenetech @NeilMBenetech can you clarify?

**Comment by johnhbenetech** [comment]
yep i agree, seems to be handled in latest commit. Thanks

**Comment by bognasoldev89** [comment]
Seems OK now, checked on Chrome Version 79.0.3945.130 and Firefox Version 72.0.2:

![781 checkbox reset](https://user-images.githubusercontent.com/51324797/74243331-dac54200-4cdf-11ea-8cb8-16088926f126.gif)

I checked found matching organizations by the way and those are filtered correctly.

**Comment by johnhbenetech** [comment]
@NeilMBenetech can you review this? I think there may have been some lack of clarity on requirements and I'm not sure it functions as you intended. @bognasoldev89 please correct me if I'm wrong but:

In the situation where I have a record in my DB which has two matches - one match that is highly similar, one match that is not so similar. This organization will not return whatsoever if the box is checked. So I wouldn't be able to view my record at all to see the highly matched partner record.

To sum it up - when this box is checked it only return your records for which ALL matches are high similarity.

**Comment by bognasoldev89** [comment]
@johnhbenetech can you provide examples?

For UWBA I found 3 exemplary organizations that show matches over and below 75%:
HOUSING AUTHORITY OF THE COUNTY OF Commoncounty (ABCD)
GLIDE MEMORIAL UNITED METHODIST CHURCH GLIDE FOUNDATION
CATHOLIC CHARITIES CYO SAN FRANCISCO, SAN MATEO & MARIN

For SMCConnect:
California State Department of Motor Vehicles (DMV)
Puente de la Costa Sur
Pacific Gas and Electric Company (PG&E)

For healthleads:
Larkin Street Youth Services
Caminar
California Department of Equal Employment and Housing

For ShelterTech:
Compass Family Services
Bay Area Women's and Children's Center (BAWCC)

For Eden:
CENTRO LEGAL DE LA RAZA
BETTER BUSINESS BUREAU (BBB)
LEGAL AID AT WORK

And for LAAC all of the matches are shown as below 75%

**Comment by kdondziak** [comment]
I implemented it in following way:
Two organizations has 'match'. Each match has similarity field. Implemented checkbox filters records to have similarity field > 75%. So, on home page there are records that have at least 1 'highly matched' organization match. On click user sees compare page. On that page user can see his/her own organization with organization matches with similarity < 75% if there are some and at least one organization with similarity > 75%.

**Comment by johnhbenetech** [comment]
gotcha, thanks

@NeilMBenetech  Can you review some matches and see if you think 75% is the right threshold?

**Comment by NeilMBenetech** [comment]
@johnhbenetech this looks good and I like @kdondziak 's approach, I was going to suggest the same.

**Comment by johnhbenetech** [comment]
@NeilMBenetech my question was more do you think "75" is the right cutoff. If you look down in the 60s and even the 50s there are still meaningful matches.

**Comment by NeilMBenetech** [comment]
Ah, thanks @johnhbenetech . We'll probably have to iterate on this and eventually may even need for users to set their own threshold. For now based on your comment shall we drop it to, say, 50% ?

**Comment by bognasoldev89** [comment]
I confirm that now it's needed a 50% match to show organizations with this feature.
